File: filesystem/src/main/scala/net/kemitix/thorp/filesystem/PathCache.scala
package net.kemitix.thorp.filesystem

import java.nio.file.{Path, Paths}
import java.time.Instant
import java.util.regex.Pattern

import scala.jdk.CollectionConverters._
import scala.jdk.OptionConverters._
import net.kemitix.thorp.domain.{Hashes, LastModified, MD5Hash}
import zio.{UIO, ZIO}

/**
  * Meta data for files in the current source, as of the last time Thorp processed this directory.
  *
  * <p>N.B. Does not include sub-directories.</p>
  */
final case class PathCache(
    data: PathCache.Data
) {
  def get(path: Path): Option[FileData] = data.get(path)
}

object PathCache {
  type Data = Map[Path, FileData]
  val fileName     = ".thorp.cache"
  val tempFileName = ".thorp.cache.tmp"

  def create(path: Path, fileData: FileData): UIO[Iterable[String]] =
    UIO(unsafeCreate(path, fileData))

  def unsafeCreate(path: Path, fileData: FileData): Set[String] = {
    fileData.hashes.keys.asScala
      .map(hashType => {
        val hash     = fileData.hashes.get(hashType).toScala
        val modified = fileData.lastModified
        val hashHash = hash.get.hash
        String.join(":",
                    hashType.label,
                    hashHash,
                    modified.at.toEpochMilli.toString,
                    path.toString)
      })
      .toSet
  }

  private val pattern =
    "^(?<hashtype>.+):(?<hash>.+):(?<modified>\\d+):(?<filename>.+)$"
  private val format = Pattern.compile(pattern)
  def fromLines(lines: Seq[String]): ZIO[Hasher, Nothing, PathCache] =
    ZIO
      .foreach(
        lines
          .map(format.matcher(_))
          .filter(_.matches())) { matcher =>
        for {
          hashType <- Hasher.typeFrom(matcher.group("hashtype"))
        } yield
          Paths.get(matcher.group("filename")) -> FileData
            .create(
              Hashes.create(hashType, MD5Hash.create(matcher.group("hash"))),
              LastModified.at(
                Instant.ofEpochMilli(matcher.group("modified").toLong))
            )
      }
      .catchAll({ _: IllegalArgumentException =>
        UIO(List.empty)
      })
      .map(list => mergeFileData(list))
      .map(map => PathCache(map))

  private def mergeFileData(
      list: List[(Path, FileData)]
  ): Data = {
    list.foldLeft(Map.empty[Path, FileData]) { (acc, pair) =>
      val (fileName, fileData) = pair
      acc.updatedWith(fileName)(
        _.map(fd => fd + fileData)
          .orElse(Some(fileData)))
    }
  }
}

File: filesystem/src/main/java/net/kemitix/thorp/filesystem/PathCache.java
package net.kemitix.thorp.filesystem;

import lombok.AccessLevel;
import lombok.RequiredArgsConstructor;
import net.kemitix.thorp.domain.*;

import java.nio.file.Path;
import java.nio.file.Paths;
import java.time.Instant;
import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

@RequiredArgsConstructor(access = AccessLevel.PRIVATE)
public class PathCache {
    private final Map<Path, FileData> data;
    public Optional<FileData> get(Path path) {
        return Optional.ofNullable(data.get(path));
    }

    public static final String fileName = ".thorp.cache";
    public static final String tempFileName = fileName + ".tmp";
    public static PathCache create(Map<Path, FileData> data) {
        return new PathCache(data);
    }
    public static Set<String> export(Path path, FileData fileData) {
        return fileData.hashes
                .keys()
                .stream()
                .map(hashType ->
                        fileData.hashes.get(hashType)
                                .map(MD5Hash::hash)
                                .map(hashHash -> String.join(":",
                                        hashType.label,
                                        hashHash,
                                        Long.toString(fileData.lastModified
                                                .at()
                                                .toEpochMilli()),
                                        path.toString()
                                        )))
                .filter(Optional::isPresent)
                .map(Optional::get)
                .collect(Collectors.toSet());
    }
    private static final String pattern =
            "^(?<hashtype>.+):(?<hash>.+):(?<modified>\\d+):(?<filename>.+)$";
    private static final Pattern format = Pattern.compile(pattern);
    public static PathCache fromLines(List<String> lines) {
        return PathCache.create(
                lines.stream()
                        .map(format::matcher)
                        .filter(Matcher::matches)
                        .map(matcher -> Tuple.create(
                                Paths.get(matcher.group("filename")),
                                FileData.create(
                                        getHashes(matcher),
                                        getModified(matcher)
                                ))).collect(Collectors.toMap(
                        tuple -> tuple.a,// keymapper - path
                        tuple -> tuple.b,// value mapper - file data
                        FileData::join)));// merge function
    }

    private static LastModified getModified(Matcher matcher) {
        return LastModified.at(
                Instant.ofEpochMilli(
                        Long.parseLong(
                                matcher.group("modified"))));
    }

    private static Hashes getHashes(Matcher matcher) {
        return Hashes.create(
                getHashtype(matcher),
                MD5Hash.create(matcher.group("hash")));
    }

    private static HashType getHashtype(Matcher matcher) {
        return HashGenerator.generatorFor(matcher.group("hashtype"))
                .hashType();
    }

}
